### PR TITLE
fix: forward provider-specific options via AI SDK providerOptions

### PR DESCRIPTION
**Pull request by nick-inkeep**
## Summary

- **Fix provider-specific options being silently dropped** — `anthropic.thinking`, `openai.reasoningEffort`, `google.thinkingConfig`, and all other provider-scoped AI SDK features were never forwarded to the AI SDK because `getGenerationParams()` spread them as top-level keys on `generateText()`/`streamText()` calls instead of nesting them under the `providerOptions` parameter
- **Add `classifyProviderOptions()` method** that separates flat providerOptions into three buckets: constructor config (skip), generation params (top-level), and provider-specific options (nested under `providerOptions`)
- **No call site changes** — all 5 affected call sites consume `prepareGenerationConfig()` via spread and automatically get the correct structure

### The Bug

```
DB providerOptions: { temperature: 0.7, anthropic: { thinking: { type: "enabled", budgetTokens: 10000 } } }
                                    ↓
getGenerationParams()  →  { temperature: 0.7, anthropic: { thinking: {...} } }   (flat)
                                    ↓
streamText({ model, temperature: 0.7, anthropic: { thinking: {...} }, ... })
                                    ↓
AI SDK destructures: { model, ..., providerOptions, ..., ...settings } = args
                                    ↓
❌ anthropic goes into ...settings → silently ignored
❌ No providerOptions key → provider features disabled
```

### The Fix

```
DB providerOptions: { temperature: 0.7, anthropic: { thinking: { type: "enabled", budgetTokens: 10000 } } }
                                    ↓
classifyProviderOptions()  →  generationParams: { temperature: 0.7 }
                              providerSpecificOptions: { anthropic: { thinking: {...} } }
                                    ↓
streamText({ model, temperature: 0.7, providerOptions: { anthropic: { thinking: {...} } }, ... })
                                    ↓
✅ providerOptions correctly forwarded to AI SDK
```

### Verified end-to-end

Tested with real Anthropic API calls (claude-sonnet-4-20250514 with extended thinking):
- **Buggy path:** `reasoning: []`, 13 output tokens (thinking silently disabled)
- **Correct path:** full reasoning chain, 194 output tokens

### Provider namespace coverage

| Provider | Namespace | Status |
|---|---|---|
| Anthropic | `anthropic` | In PROVIDER_NAMES |
| OpenAI | `openai` | In PROVIDER_NAMES |
| Google | `google` | In PROVIDER_NAMES |
| Azure | `azure` | In PROVIDER_NAMES (uses `openai` namespace in AI SDK) |
| Gateway | `gateway` | In PROVIDER_NAMES |
| OpenRouter | `openrouter` | In PROVIDER_NAMES |
| NIM | — | Excluded (constructor config, no providerOptions namespace) |
| Custom | — | Excluded (constructor config, no providerOptions namespace) |

## Test plan

- [x] 24 new unit tests covering all provider types, edge cases, backward compatibility
- [x] All 51 model-factory tests pass (27 existing + 24 new)
- [x] All 1621 agents-core tests pass
- [x] Typecheck passes
- [x] Lint passes
- [x] Existing `getGenerationParams` tests pass via deprecated passthrough
- [x] Existing `prepareGenerationConfig` Azure tests pass unchanged

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: packages/agents-core/src/utils/model-factory.ts [hunk 1]
```diff
@@ -217,6 +217,36 @@ export class ModelFactory {
...
+   * and the value is an object, it goes into providerOptions
+   */
+  private static readonly PROVIDER_NAMES = new Set([
+    'anthropic',
+    'openai',
+    'google',
+    'azure',
+    'gateway',
+    'openrouter',
+  ]);
```
[the remark this answers is not in the document]
Comment (nick-inkeep) [reply packages/agents-core/src/utils/model-factory.ts:266]: Fixed — removed `gateway` from `PROVIDER_NAMES` and added it to the skip branch alongside `nim`/`custom`. Tests updated.

File: packages/agents-core/src/utils/model-factory.ts [hunk 2]
```diff
@@ -261,44 +291,67 @@ export class ModelFactory {
...
+        // Skip — handled by extractProviderConfig()
+      } else if (key === 'maxDuration') {
+        maxDuration = value as number;
+      } else if (ModelFactory.META_KEYS.has(key)) {
+        // Skip — extracted but not forwarded (e.g., contextWindowSize)
+      } else if (
+        ModelFactory.PROVIDER_NAMES.has(key) &&
+        typeof value === 'object' &&
+        value !== null
+      ) {
```
[the remark this answers is not in the document]
Comment (nick-inkeep) [reply packages/agents-core/src/utils/model-factory.ts:347]: Fixed — added `\!Array.isArray(value)` guard. Also added a test case for this.

File: packages/agents-core/src/utils/model-factory.ts [hunk 3]
```diff
@@ -217,6 +217,36 @@ export class ModelFactory {
...
+    'baseUrl',
+    'headers',
+    'resourceName',
+    'apiVersion',
+  ]);
+
+  /**
+   * Keys that are extracted but not forwarded as generation params
+   */
+  private static readonly META_KEYS = new Set(['maxDuration', 'contextWindowSize']);
```
[the remark this answers is not in the document]
Comment (nick-inkeep) [reply packages/agents-core/src/utils/model-factory.ts]: Fixed — removed `maxDuration` from `META_KEYS`, leaving only `contextWindowSize`.

**Comment by nick-inkeep** [comment]
Closing — the upstream fix in #2523 (extractStreamProviderOptions heuristic) is correct for all real-world inputs. Our explicit classifyProviderOptions approach introduced a gateway regression by incorrectly skipping gateway provider options. The heuristic handles all documented provider namespaces correctly and the only theoretical imprecision (forwarding unknown object-valued keys) is harmless since the AI SDK ignores unknown provider namespaces.
